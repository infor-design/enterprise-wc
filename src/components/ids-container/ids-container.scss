/* Ids Tag Component */
@import '../../core/ids-base';

// Host styles are needed here for "placed" components like Popup.
// (acts as a CSS "reset" without needing a real one)
:host {
  @include block();

  contain: content;
  height: 100vh;
  width: 100vw;
}

:host([hidden]) {
  display: none;
}

.ids-container {
  @include basic-color-transition();
  @include outline-none();
  @include p-0();

  box-sizing: border-box;
  height: 100%;
  overflow: auto;
  scroll-behavior: smooth;
  width: 100%;

  &[scrollable='false'] {
    overflow: hidden;
  }
}

// Handle Themes
.ids-container[mode='light'],
.ids-container:not([mode]) {
  @include bg-white();
  @include text-slate-100();

  &.color-variant-alternate {
    @include bg-slate-70();
    @include text-slate-10();
  }
}

.ids-container[mode='dark'] {
  @include bg-slate-90();
  @include text-slate-10();

  &.color-variant-alternate {
    @include bg-black();
    @include text-slate-10();
  }
}

.ids-container[mode='contrast'] {
  @include bg-slate-10();
  @include text-slate-100();

  &.color-variant-alternate {
    @include bg-black();
    @include text-white();
  }
}

.ids-container[version='classic'][mode='light'] {
  @include bg-graphite-10();
  @include text-graphite-60();

  &.color-variant-alternate {
    @include bg-graphite-70();
    @include text-graphite-20();
  }
}

.ids-container[version='classic'][mode='dark'] {
  @include bg-classic-slate-90();
  @include text-classic-slate-10();

  &.color-variant-alternate {
    @include bg-graphite-90();
    @include text-classic-slate-10();
  }
}

.ids-container[version='classic'][mode='contrast'] {
  @include bg-graphite-20();
  @include text-graphite-80();

  &.color-variant-alternate {
    @include bg-graphite-90();
    @include text-graphite-20();
  }
}
