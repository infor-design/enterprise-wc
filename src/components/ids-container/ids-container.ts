import { customElement, scss } from '../../core/ids-decorators';
import { attributes } from '../../core/ids-attributes';
import { stringToBool } from '../../utils/ids-string-utils/ids-string-utils';

import Base from './ids-container-base';
import locale from '../ids-locale/ids-locale-global';

import styles from './ids-container.scss';

/**
 * IDS Container Component
 * @type {IdsContainer}
 * @inherits IdsElement
 * @mixes IdsThemeMixin
 * @mixes IdsEventsMixin
 * @mixes IdsLocaleMixin
 * @part container - the entire container element
 */
@customElement('ids-container')
@scss(styles)
export default class IdsContainer extends Base {
  constructor() {
    super();
    this.state.locale = locale;
  }

  /**
   * Invoked each time the custom element is appended into a document-connected element.
   */
  connectedCallback() {
    super.connectedCallback();
    if (this.reset) {
      this.#addReset();
    }

    // Set initial lang and locale
    this.setAttribute('language', this.state.locale.state.language);
    this.setAttribute('locale', this.state.locale.state.localeName);

    // Remove hidden for FOUC
    this.onEvent('load.container', window, () => {
      this.removeAttribute('hidden');
      this.offEvent('load.container', window);
    });

    // In some cases the page may be loaded
    if (document.readyState === 'complete') {
      this.removeAttribute('hidden');
    }
    if (this.padding) this.container.style.padding = `${this.padding}px`;
  }

  /**
   * Invoked each time the custom element is removed from a document-connected element.
   */
  disconnectedCallback() {
    super.disconnectedCallback();
    this.state = undefined;
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array} The attributes in an array
   */
  static get attributes() {
    return [
      ...super.attributes,
      attributes.LANGUAGE,
      attributes.LOCALE,
      attributes.PADDING,
      attributes.RESET,
      attributes.SCROLLABLE,
      attributes.MODE
    ];
  }

  /**
   * Create the Template for the contents
   * @returns {string} The template
   */
  template(): string {
    return `<div class="ids-container" part="container"${this.scrollable === 'true' ? ' tabindex="0"' : ''}><slot></slot></div>`;
  }

  /**
   * Inherited from `IdsColorVariantMixin`
   * @returns {Array<string>} List of available color variants for this component
   */
  colorVariants = ['alternate'];

  /**
   * If set to number the container will have padding added (in pixels)
   * @param {string} value sets the padding to the container
   */
  set padding(value: string) {
    if (this.container) this.container.style.padding = `${value}px`;
    this.setAttribute(attributes.PADDING, value.toString());
  }

  get padding(): string {
    return this.getAttribute(attributes.PADDING);
  }

  /**
   * If set to true the container is scrollable
   * @param {boolean|string} value true of false depending if the tag is scrollable
   */
  set scrollable(value: boolean | string) {
    if (stringToBool(value)) {
      this.setAttribute(attributes.SCROLLABLE, 'true');
      this.container.setAttribute(attributes.SCROLLABLE, 'true');
      this.container.setAttribute('tabindex', '0');
      return;
    }

    this.setAttribute(attributes.SCROLLABLE, 'false');
    this.container.setAttribute(attributes.SCROLLABLE, 'false');
    this.container.removeAttribute('tabindex');
  }

  get scrollable(): boolean | string { return this.getAttribute(attributes.SCROLLABLE) || 'true'; }

  /**
   * Add the reset to the body
   * @private
   */
  #addReset() {
    const body = document.querySelector('body');
    if (body) {
      body.style.margin = '0';
    }
  }

  /**
   * If set to true body element will get reset
   * @param {boolean|string} value true of false
   */
  set reset(value: boolean | string) {
    if (stringToBool(value)) {
      this.#addReset();
      return;
    }
    this.removeAttribute(attributes.RESET);

    const body = document.querySelector('body');
    if (body) {
      body.style.margin = '';
    }
  }

  get reset(): boolean | string { return this.getAttribute(attributes.RESET) || 'true'; }

  /**
   * Set the language for a component and wait for it to finish (async)
   * @param {string} value The language string value
   */
  async setLanguage(value: string) {
    await this.state.locale.setLanguage(value);
    this.language = value;
    this.triggerEvent('languagechange', this, { detail: { elem: this, language: this.language, locale: this.state?.locale } });
  }

  /**
   * Set the language for a component
   * @param {string} value The language string value
   */
  set language(value: string) {
    if (value) {
      this.state.locale.setLanguage(value);
      this.state.locale.updateLangTag(this, value);
      this.setAttribute('language', value);
      requestAnimationFrame(() => {
        this.triggerEvent('languagechange', this, { detail: { elem: this, language: this.language, locale: this.state?.locale } });
      });
    }
  }

  /**
   * Get the language data keys and message for the current language
   * @returns {object} The language data object
   */
  get language(): string {
    return this.state?.locale?.language;
  }

  /**
   * Set the locale for a component and wait for it to finish (async)
   * @param {string} value The locale string value
   */
  async setLocale(value: string) {
    if (value) {
      await this.state.locale.setLocale(value);
      const lang = this.state.locale.correctLanguage(value);
      this.setAttribute('locale', value);
      this.setAttribute('language', lang);
      this.state.locale.updateLangTag(this, lang);
      this.triggerEvent('localechange', this, { detail: { elem: this, language: this.language, locale: this.state?.locale } });
    }
  }

  /**
   * Set the locale for a component
   * @param {string} value The locale string value
   */
  set locale(value: string) {
    if (value) {
      this.state.locale.setLocale(value);
      const lang = this.state.locale.correctLanguage(value);
      this.setAttribute('locale', value);
      this.setAttribute('language', lang);
      this.state.locale.updateLangTag(this, lang);

      requestAnimationFrame(() => {
        this.triggerEvent('localechange', this, { detail: { elem: this, language: this.language, locale: this.state?.locale } });
      });
    }
  }

  get locale(): string {
    return this.state.locale;
  }

  get localeName(): string {
    return this.state.locale.state.localeName;
  }
}
