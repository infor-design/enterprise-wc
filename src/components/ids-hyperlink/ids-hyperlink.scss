/* Ids Hyperlink Css */
@import '../../core/ids-base';

/* Ids Typography System */
.ids-text-10 {
  @include text-10();
  @include text-uppercase();
}

.ids-text-12,
.ids-text-xs {
  @include text-12();
  @include text-uppercase();
}

.ids-text-14,
.ids-text-sm {
  @include text-14();
}

.ids-text-16,
.ids-text-base {
  @include text-16();
}

.ids-text-20 {
  @include text-20();
}

.ids-text-24,
.ids-text-lg {
  @include text-24();
}

.ids-text-28 {
  @include text-28();
}

.ids-text-32,
.ids-text-xl {
  @include text-32();
}

.ids-text-40 {
  @include text-40();
}

.ids-text-48 {
  @include text-48();
}

.ids-text-60 {
  @include text-60();
}

.ids-text-72 {
  @include text-72();
  @include mb-0();
}

.ids-hyperlink {
  @include antialiased();
  @include cursor-pointer();
  @include font-sans();
  @include font-normal();
  @include p-4();
  @include rounded-default();
  @include mb-8();
  @include relative();

  border: 1px solid transparent;
  display: inline-block;
  margin: -4px;

  &[disabled] {
    @include cursor-default();

    pointer-events: none;
    user-select: none;
  }

  &:focus {
    outline: none;
  }

  &.bold {
    @include font-bold();
  }

  &.hitbox::after {
    content: '';
    display: block;
    height: 44px;
    top: 50%;
    left: 0;
    position: absolute;
    transform: translateY(-50%);
    width: 100%;
  }

  // =========================================
  // Theme colors are applied when `color="unset"` is not set
  &:not(.ids-hyperlink-color-unset) {
    &[version='new'][mode='light'] {
      @include text-azure-60();

      &:focus {
        border: 1px solid var(--ids-color-palette-azure-60);
        box-shadow: 0 0 4px 3px rgb(54 138 192 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-70();
      }

      &[disabled] {
        @include text-slate-40();
      }
    }

    &[version='new'][mode='dark'] {
      @include text-azure-40();

      &:focus {
        border: 1px solid var(--ids-color-palette-azure-60);
        box-shadow: 0 0 4px 2px rgb(141 201 230 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-60();
      }

      &[disabled] {
        @include text-slate-50();
      }
    }

    &[version='new'][mode='contrast'] {
      @include text-azure-90();

      &:focus {
        border: 1px solid var(--ids-color-palette-azure-90);
        box-shadow: 0 0 4px 2px rgb(41 41 41 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-100();
      }

      &[disabled] {
        @include text-slate-50();
      }
    }

    &[version='classic'][mode='light'] {
      @include text-azure-70();

      &:focus {
        box-shadow: 0 0 4px 3px rgb(54 138 192 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-80();
      }

      &[disabled] {
        @include text-graphite-40();
      }
    }

    &[version='classic'][mode='dark'] {
      @include text-azure-40();

      &:focus {
        box-shadow: 0 0 4px 3px rgb(141 201 230 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-50();
      }

      &[disabled] {
        @include text-classic-slate-50();
      }
    }

    &[version='classic'][mode='contrast'] {
      @include text-azure-90();

      &:focus {
        border: 1px solid var(--ids-color-palette-azure-90);
        box-shadow: 0 0 4px 2px rgb(41 41 41 / 0.3);
      }

      &:hover:not([disabled]):not(.ids-text-decoration-hover),
      &:active:not([disabled]):not(.ids-text-decoration-hover) {
        @include text-azure-100();
      }

      &[disabled] {
        @include text-classic-slate-50();
      }
    }

    // ===============================================
    // Color Variants
    &.color-variant-breadcrumb {
      &[version='new'][mode='light'] {
        @include text-slate-60();

        &[disabled] {
          @include text-slate-30();
        }
      }

      &[version='new'][mode='dark'] {
        @include text-slate-40();

        &[disabled] {
          @include text-slate-60();
        }
      }

      &[version='new'][mode='contrast'] {
        @include text-slate-80();

        &[disabled] {
          @include text-slate-30();
        }
      }

      &[version='classic'][mode='light'] {
        @include text-slate-60();

        &[disabled] {
          @include text-slate-30();
        }
      }

      &[version='classic'][mode='dark'] {
        @include text-slate-40();

        &[disabled] {
          @include text-classic-slate-50();
        }
      }

      &[version='classic'][mode='contrast'] {
        @include text-slate-80();

        &[disabled] {
          @include text-slate-50();
        }
      }

      &:not([mode]) {
        @include text-slate-60();
      }
    }

    &.color-variant-alternate {
      &[disabled] {
        opacity: 0.7;
      }

      &[version='new'][mode='light'] {
        @include text-white();
      }

      &[version='new'][mode='dark'] {
        @include text-white();
      }

      &[version='new'][mode='contrast'] {
        @include text-white();
      }

      &[version='classic'][mode='light'] {
        @include text-white();
      }

      &[version='classic'][mode='dark'] {
        @include text-white();
      }

      &[version='classic'][mode='contrast'] {
        @include text-white();
      }
    }
  }
}

.ids-text-decoration-none {
  text-decoration: none;

  &:hover {
    text-decoration: none;
  }
}

.ids-text-decoration-hover {
  text-decoration: none;

  &[href]:hover {
    text-decoration: underline;
  }
}
