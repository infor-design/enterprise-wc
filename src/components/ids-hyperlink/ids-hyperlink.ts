import fontSizes from 'ids-identity/dist/theme-new/tokens/web/ui.config.font-sizes';
import { customElement, scss } from '../../core/ids-decorators';
import { attributes } from '../../core/ids-attributes';
import Base from './ids-hyperlink-base';
import { stringToBool } from '../../utils/ids-string-utils/ids-string-utils';

import styles from './ids-hyperlink.scss';

/**
 * IDS Hyperlink Component
 * @type {IdsHyperlink}
 * @inherits IdsElement
 * @mixes IdsHitboxMixin
 * @mixes IdsThemeMixin
 * @mixes IdsEventsMixin
 * @part link - the link element
 */
@customElement('ids-hyperlink')
@scss(styles)
export default class IdsHyperlink extends Base {
  constructor() {
    super();
  }

  connectedCallback() {
    if (!this.getAttribute('role')) {
      this.setAttribute('role', 'link');
    }
    super.connectedCallback();
    if (!(this.getAttribute('role'))) this.setAttribute('role', 'link');
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array} The attributes in an array
   */
  static get attributes() {
    return [
      ...super.attributes,
      attributes.COLOR,
      attributes.DISABLED,
      attributes.HREF,
      attributes.FONT_SIZE,
      attributes.FONT_WEIGHT,
      attributes.MODE,
      attributes.TARGET,
      attributes.TEXT_DECORATION,
      attributes.VERSION
    ];
  }

  /**
   * Inner template contents
   * @returns {string} The template
   */
  template(): string {
    return `<a class="ids-hyperlink" part="link"><slot></slot></a>`;
  }

  /**
   *
   */
  colorVariants = ['breadcrumb', 'alternate'];

  /**
   * Set the link href
   * @param {string} value Set the link's href to some link
   */
  set href(value: string) {
    if (value) {
      this.setAttribute(attributes.HREF, value);
      this.container.setAttribute(attributes.HREF, value);
      return;
    }
    this.removeAttribute(attributes.HREF);
    this.container.removeAttribute(attributes.HREF);
  }

  get href(): string { return this.getAttribute(attributes.HREF); }

  /**
   * Set the link target attribute
   * @param {string} value Set the link's href to some link
   */
  set target(value: string) {
    if (value) {
      this.setAttribute(attributes.TARGET, value);
      this.container.setAttribute(attributes.TARGET, value);
      return;
    }
    this.removeAttribute(attributes.TARGET);
    this.container.removeAttribute(attributes.TARGET);
  }

  get target(): string { return this.getAttribute(attributes.TARGET); }

  /**
   * Set the link text decoration styling
   * @param {string} value If 'none', removes text decoration, If hover then just on hover it
   * is shown.
   */
  set textDecoration(value: string) {
    if (value?.toLowerCase() === 'none') {
      this.setAttribute(attributes.TEXT_DECORATION, value);
      this.container.classList.add('ids-text-decoration-none');
      return;
    }
    if (value?.toLowerCase() === 'hover') {
      this.setAttribute(attributes.TEXT_DECORATION, value);
      this.container.classList.add('ids-text-decoration-hover');
      return;
    }
    this.removeAttribute(attributes.TEXT_DECORATION);
    this.container.classList.remove('ids-text-decoration-none');
    this.container.classList.remove('ids-text-decoration-hover');
  }

  get textDecoration(): string { return this.getAttribute(attributes.TEXT_DECORATION); }

  /**
   * Set the text to disabled color.
   * @param {boolean} value True if disabled
   */
  set disabled(value: boolean) {
    const val = stringToBool(value);
    if (val) {
      this.setAttribute(attributes.DISABLED, `${value}`);
      this.container.setAttribute(attributes.DISABLED, `${value}`);
      this.container.setAttribute('tabindex', '-1');
      return;
    }
    this.removeAttribute(attributes.DISABLED);
    this.container.removeAttribute(attributes.DISABLED);
    this.container.removeAttribute('tabindex');
  }

  get disabled(): boolean { return this.getAttribute(attributes.DISABLED); }

  /**
   *
   * If set to "unset", color can be controlled by parent container
   * @param {string | null} value  "unset" or undefined/null
   */
  set color(value: string | null) {
    if (value === 'unset') {
      this.setAttribute(attributes.COLOR, value);
      this.container.classList.add('ids-hyperlink-color-unset');
    } else {
      this.removeAttribute(attributes.COLOR);
      this.container.classList.remove('ids-hyperlink-color-unset');
    }
  }

  get color(): string | null {
    return this.getAttribute(attributes.COLOR);
  }

  /**
   * Set the font size/style of the text with a class.
   * @param {string | null} value The font size in the font scheme
   * i.e. 10, 12, 16 or xs, sm, base, lg, xl
   */
  set fontSize(value: string | null) {
    fontSizes?.forEach((size: string) => {
      this.container?.classList.remove(`ids-text-${Object.keys(size)}`);
    });

    if (value) {
      this.setAttribute(attributes.FONT_SIZE, value);
      this.container?.classList.add(`ids-text-${value}`);
      return;
    }

    this.removeAttribute(attributes.FONT_SIZE);
  }

  get fontSize(): string | null { return this.getAttribute(attributes.FONT_SIZE); }

  /**
   * Adjust font weight; can be either "bold" "lighter" or not present
   * since font supports 300, 400, 600
   * @param {string | null} value (if bold)
   */
  set fontWeight(value: string | null) {
    this.container?.classList.remove('bold', 'lighter');

    if (value === 'bold' || value === 'lighter') {
      this.setAttribute(attributes.FONT_WEIGHT, value);
      this.container?.classList.add(value);
      return;
    }

    this.removeAttribute(attributes.FONT_WEIGHT);
  }

  get fontWeight(): string | null {
    return this.getAttribute(attributes.FONT_WEIGHT);
  }
}
