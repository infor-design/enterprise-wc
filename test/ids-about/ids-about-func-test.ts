/**
 * @jest-environment jsdom
 */
import '../helpers/resize-observer-mock';
import '../helpers/match-media-mock';
import processAnimFrame from '../helpers/process-anim-frame';

import IdsAbout from '../../src/components/ids-about/ids-about';

// Supporing components
import '../../src/components/ids-container/ids-container';
import '../../src/components/ids-text/ids-text';
import '../../src/components/ids-hyperlink/ids-hyperlink';

// locale messages
import { messages as esMessages } from '../../src/components/ids-locale/cultures/es-messages';
import { messages as jaMessages } from '../../src/components/ids-locale/cultures/ja-messages';
import testForXssVulnerabilities from '../helpers/xss-helper';

const name = 'ids-about';
const id = 'test-about-component';
const productVersion = '4.0.0';
const productName = 'Controls';
const copyrightYear = '2022';
const deviceSpecs = true;
const useDefaultCopyright = true;

describe('IdsAbout Component (using properties)', () => {
  let component: any;

  beforeEach(async () => {
    const container: any = document.createElement('ids-container');
    document.body.appendChild(container);
    await container.setLanguage('en');

    component = new IdsAbout();
    component.id = id;
    component.productVersion = productVersion;
    component.productName = productName;
    component.copyrightYear = copyrightYear;
    component.deviceSpecs = deviceSpecs;
    component.useDefaultCopyright = useDefaultCopyright;
    container.appendChild(component);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    component = null;
  });

  it('should render', () => {
    const errors = jest.spyOn(global.console, 'error');

    expect(document.querySelectorAll(name).length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();

    // Use Snapshots
    expect(component.outerHTML).toMatchSnapshot();
    component.show();
    expect(component.outerHTML).toMatchSnapshot();

    component.hide();
    expect(component.outerHTML).toMatchSnapshot();
  });

  it('can be destroyed', () => {
    const errors = jest.spyOn(global.console, 'error');

    component.remove();

    expect(document.querySelectorAll(name).length).toEqual(0);
    expect(errors).not.toHaveBeenCalled();
  });

  it('has properties', () => {
    expect(component.productVersion).toEqual(productVersion);
    expect(component.productName).toEqual(productName);
    expect(component.copyrightYear).toEqual(copyrightYear);
    expect(component.deviceSpecs).toEqual(deviceSpecs);
    expect(component.useDefaultCopyright).toEqual(useDefaultCopyright);
  });

  it('can alter productVersion and productName', () => {
    const newProductVersion = 'New productVersion';
    const newProductName = 'New productVersion';
    component.productVersion = newProductVersion;
    component.productName = newProductName;

    expect(component.productVersion).toEqual(newProductVersion);
    expect(component.productName).toEqual(newProductName);

    component.productVersion = '';
    component.productName = '';

    expect(component.productVersion).toEqual('');
    expect(component.productName).toEqual('');

    testForXssVulnerabilities(component, 'productName', '');
    testForXssVulnerabilities(component, 'productVersion', '');
  });

  it('can alter useDefaultCopyright', () => {
    component.useDefaultCopyright = false;

    expect(component.useDefaultCopyright).toBeFalsy();

    component.useDefaultCopyright = undefined;

    expect(component.useDefaultCopyright).toBeFalsy();

    component.useDefaultCopyright = 'false';

    expect(component.useDefaultCopyright).toBeFalsy();

    component.useDefaultCopyright = 'true';

    expect(component.useDefaultCopyright).toBeTruthy();
  });

  it('can alter deviceSpecs', () => {
    component.deviceSpecs = false;

    expect(component.deviceSpecs).toBeFalsy();

    component.deviceSpecs = undefined;

    expect(component.deviceSpecs).toBeFalsy();

    component.deviceSpecs = 'false';

    expect(component.deviceSpecs).toBeFalsy();

    component.deviceSpecs = 'true';

    expect(component.deviceSpecs).toBeTruthy();
  });

  it('can alter copyrightYear', () => {
    component.copyrightYear = 2015;

    expect(component.copyrightYear).toEqual('2015');
  });
});

describe('IdsAbout Component (using attributes)', () => {
  let component: any;

  beforeEach(async () => {
    document.body.insertAdjacentHTML('beforeend', `<ids-about id="about-example" product-name="Controls" product-version="4.0.0">
      <ids-icon slot="icon" icon="logo" viewbox="0 0 35 34" size="large" /></ids-icon>
      <ids-text id="about-example-name" slot="appName" type="h1" font-size="24">IDS Enterprise</ids-text>
      <ids-text id="about-example-content" slot="content" type="p">Fashionable components for fashionable applications.</ids-text>
    </ids-about>`);
    component = document.querySelector(name);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    component = null;
  });

  it('should render', () => {
    const errors = jest.spyOn(global.console, 'error');

    expect(document.querySelectorAll(name).length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('can be destroyed', () => {
    const errors = jest.spyOn(global.console, 'error');

    component.remove();

    expect(document.querySelectorAll(name).length).toEqual(0);
    expect(errors).not.toHaveBeenCalled();
  });

  it('has properties', () => {
    expect(component.productName).toEqual(productName);
    expect(component.productVersion).toEqual(productVersion);
    expect(component.copyrightYear).toEqual(new Date().getFullYear());
    expect(component.useDefaultCopyright).toBeTruthy();
    expect(component.deviceSpecs).toBeTruthy();
  });

  it('should not close on click outside', (done) => {
    const clickEvent = new MouseEvent('click', { bubbles: true });

    component.visible = true;
    setTimeout(() => {
      // Click outside the about into the overlay area
      document.body.dispatchEvent(clickEvent);
      // Nor should calling the method directly
      component.onOutsideClick();

      setTimeout(() => {
        expect(component.visible).toEqual(true);
        done();
      });
    }, 70);
  });
});

describe('IdsAbout Component (empty)', () => {
  let component: any;

  beforeEach(async () => {
    document.body.insertAdjacentHTML('beforeend', `<ids-about id="${id}"></ids-about>`);
    component = document.querySelector(name);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    component = null;
  });

  it('should render', () => {
    const errors = jest.spyOn(global.console, 'error');

    expect(document.querySelectorAll(name).length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('should have default properties', () => {
    expect(component.copyrightYear).toEqual(new Date().getFullYear());
    expect(component.deviceSpecs).toBeTruthy();
    expect(component.useDefaultCopyright).toBeTruthy();
  });

  it('can change productVersion and productName after being invoked', () => {
    const newProductVersion = '1.0.1';
    const newProductName = 'IdsAbout';

    component.productVersion = newProductVersion;
    component.productName = newProductName;

    expect(component.querySelectorAll('*').length).toBeTruthy();
    expect(component.productName).toEqual(newProductName);
    expect(component.productVersion).toEqual(newProductVersion);
  });

  it('can change copyrightYear after being invoked', () => {
    component.copyrightYear = 2015;

    expect(component.querySelectorAll('*').length).toBeTruthy();
    expect(component.copyrightYear).toEqual('2015');
  });

  it('can remove default copyright after being invoked', () => {
    component.useDefaultCopyright = false;

    expect(component.useDefaultCopyright).toBeFalsy();
  });

  it('can remove device specs after being invoked', () => {
    component.deviceSpecs = false;

    expect(component.deviceSpecs).toBeFalsy();
  });

  it('can click outside and it wont close', (done) => {
    const clickEvent = new MouseEvent('click', { bubbles: true });

    component.visible = true;
    setTimeout(() => {
      document.body.dispatchEvent(clickEvent);
      component.onOutsideClick();

      setTimeout(() => {
        expect(component.visible).toEqual(true);
        done();
      });
    }, 70);
  });
});

describe('IdsAbout Component locale', () => {
  let component: any;
  let container: any;

  beforeEach(async () => {
    container = document.createElement('ids-container');
    document.body.appendChild(container);
    await container.setLanguage('en');

    component = new IdsAbout();
    component.id = id;
    component.productVersion = productVersion;
    component.productName = productName;
    component.copyrightYear = copyrightYear;
    component.deviceSpecs = deviceSpecs;
    component.useDefaultCopyright = useDefaultCopyright;
    container.appendChild(component);
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    component = null;
    container = null;
  });

  it('can change language from the container', async () => {
    await container.setLanguage('ar');
    expect(container.getAttribute('dir')).toEqual('rtl');

    await container.setLanguage('es');
    await processAnimFrame();

    // if copyright translates
    let copyrightOriginal = esMessages.AboutText.value
      .replace('&copy; {0}', `© ${copyrightYear}`)
      .concat(' www.infor.com.');
    let copyrightReceived = document.querySelector('[slot="copyright"]')?.textContent;
    expect(copyrightOriginal).toEqual(copyrightReceived);

    // if device specs translates
    let deviceSpecsText = document.querySelector('[slot="device"]')?.textContent;

    expect(deviceSpecsText).toContain(esMessages.Platform.value);

    await container.setLanguage('ja');
    await processAnimFrame();

    // if copyright translates
    copyrightOriginal = jaMessages.AboutText.value
      .replace('&copy; {0}', `© ${copyrightYear}`)
      .concat(' www.infor.com.');
    copyrightReceived = document.querySelector('[slot="copyright"]')?.textContent;

    expect(copyrightOriginal).toEqual(copyrightReceived);

    // if device specs translates
    deviceSpecsText = document.querySelector('[slot="device"]')?.textContent;

    expect(deviceSpecsText).toContain(jaMessages.Platform.value);
  });
});
