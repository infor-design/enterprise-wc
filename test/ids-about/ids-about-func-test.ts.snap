// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsAbout Component (using properties) should render 1`] = `
"<ids-about id=\\"test-about-component\\" product-version=\\"4.0.0\\" product-name=\\"Controls\\" copyright-year=\\"2022\\" device-specs=\\"true\\" use-default-copyright=\\"true\\" aria-label=\\" Controls 4.0.0\\" role=\\"dialog\\"><ids-text slot=\\"product\\" type=\\"p\\">Controls 4.0.0</ids-text><ids-modal-button cancel=\\"\\" slot=\\"buttons\\" type=\\"tertiary\\" css-class=\\"ids-icon-button ids-modal-icon-button\\"><span class=\\"audible\\">Close modal</span><ids-icon slot=\\"icon\\" icon=\\"close\\"></ids-icon></ids-modal-button><ids-text slot=\\"device\\" type=\\"p\\">
        <span>Platform : </span><br>
        <span>Mobile : false</span><br>
        <span>Browser :  jsdom (19.0.0)</span><br>
        <span>Locale : en-US</span><br>
        <span>Language : en</span><br>
        <span>Browser Language : en-US</span><br>
        <span>IDS Version : 0.0.0-beta.22</span>
      </ids-text><ids-text slot=\\"copyright\\" type=\\"p\\">Copyright © 2022 Infor. All rights reserved. The word and design marks set forth herein are trademarks and/or registered trademarks of Infor and/or its affiliates and subsidiaries. All other trademarks listed herein are the property of their respective owners <ids-hyperlink target=\\"_blank\\" href=\\"https://www.infor.com\\" role=\\"link\\">www.infor.com</ids-hyperlink>.</ids-text></ids-about>"
`;

exports[`IdsAbout Component (using properties) should render 2`] = `
"<ids-about id=\\"test-about-component\\" product-version=\\"4.0.0\\" product-name=\\"Controls\\" copyright-year=\\"2022\\" device-specs=\\"true\\" use-default-copyright=\\"true\\" aria-label=\\" Controls 4.0.0\\" role=\\"dialog\\" visible=\\"\\" style=\\"z-index: 1020;\\" captures-focus=\\"true\\"><ids-text slot=\\"product\\" type=\\"p\\">Controls 4.0.0</ids-text><ids-modal-button cancel=\\"\\" slot=\\"buttons\\" type=\\"tertiary\\" css-class=\\"ids-icon-button ids-modal-icon-button\\"><span class=\\"audible\\">Close modal</span><ids-icon slot=\\"icon\\" icon=\\"close\\"></ids-icon></ids-modal-button><ids-text slot=\\"device\\" type=\\"p\\">
        <span>Platform : </span><br>
        <span>Mobile : false</span><br>
        <span>Browser :  jsdom (19.0.0)</span><br>
        <span>Locale : en-US</span><br>
        <span>Language : en</span><br>
        <span>Browser Language : en-US</span><br>
        <span>IDS Version : 0.0.0-beta.22</span>
      </ids-text><ids-text slot=\\"copyright\\" type=\\"p\\">Copyright © 2022 Infor. All rights reserved. The word and design marks set forth herein are trademarks and/or registered trademarks of Infor and/or its affiliates and subsidiaries. All other trademarks listed herein are the property of their respective owners <ids-hyperlink target=\\"_blank\\" href=\\"https://www.infor.com\\" role=\\"link\\">www.infor.com</ids-hyperlink>.</ids-text></ids-about>"
`;

exports[`IdsAbout Component (using properties) should render 3`] = `
"<ids-about id=\\"test-about-component\\" product-version=\\"4.0.0\\" product-name=\\"Controls\\" copyright-year=\\"2022\\" device-specs=\\"true\\" use-default-copyright=\\"true\\" aria-label=\\" Controls 4.0.0\\" role=\\"dialog\\" style=\\"z-index: 1020;\\" captures-focus=\\"true\\"><ids-text slot=\\"product\\" type=\\"p\\">Controls 4.0.0</ids-text><ids-modal-button cancel=\\"\\" slot=\\"buttons\\" type=\\"tertiary\\" css-class=\\"ids-icon-button ids-modal-icon-button\\"><span class=\\"audible\\">Close modal</span><ids-icon slot=\\"icon\\" icon=\\"close\\"></ids-icon></ids-modal-button><ids-text slot=\\"device\\" type=\\"p\\">
        <span>Platform : </span><br>
        <span>Mobile : false</span><br>
        <span>Browser :  jsdom (19.0.0)</span><br>
        <span>Locale : en-US</span><br>
        <span>Language : en</span><br>
        <span>Browser Language : en-US</span><br>
        <span>IDS Version : 0.0.0-beta.22</span>
      </ids-text><ids-text slot=\\"copyright\\" type=\\"p\\">Copyright © 2022 Infor. All rights reserved. The word and design marks set forth herein are trademarks and/or registered trademarks of Infor and/or its affiliates and subsidiaries. All other trademarks listed herein are the property of their respective owners <ids-hyperlink target=\\"_blank\\" href=\\"https://www.infor.com\\" role=\\"link\\">www.infor.com</ids-hyperlink>.</ids-text></ids-about>"
`;
